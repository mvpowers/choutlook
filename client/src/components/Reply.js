import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import {
  FaBold,
  FaItalic,
  FaUnderline,
  FaAlignCenter,
  FaAlignJustify,
  FaAlignLeft,
  FaAlignRight,
  FaListUl,
  FaListOl,
} from 'react-icons/lib/fa';

const Container = styled.div`
  background-color: #fff;
  border: 1px solid #eaeaea;
  margin: 10px;
  flex: 1;
`;

const Sender = styled.div`
  display: flex;
  flex-direction: row;
  padding: 10px;
  font-size: 12px;
  align-items: center;
`;

const Bubble = styled.div`
  display: inline-block;
  border-radius: 100px;
  background-color: #1e7145;
  color: #fff;
  padding: 10px 10px;
  margin: 10px;
`;

const Text = styled.textarea`
  width: 90%;
  height: 100px;
  font-size: 12px;
  border: none;
  border-radius: 5px;
  margin: 0 20px;
  padding: 10px;
  box-sizing: border-box;
  resize: none;
`;

const Tools = styled.div`
  display: flex;
  flex-direction: row;
  margin: 0 20px;
`;

const Icon = styled.div`
  padding: 10px;
  color: #666;

  &:hover {
    background-color: #eaeaea;
  }
`;

const Send = styled.div`
  background-color: #0078d7;
  color: #fff;
  padding: 8px 18px;
  margin: 20px;
  border-radius: 2px;
  font-size: 12px;
  font-weight: 900;
  display: inline-block;

  &:hover {
    background-color: #106ebe;
  }
`;

const Discard = styled.div`
  color: #333;
  padding: 8px 18px;
  margin: 20px 0;
  border-radius: 2px;
  font-size: 12px;
  font-weight: 900;
  display: inline-block;

  &:hover {
    background-color: #dadada;
  }
`;

const NameInput = styled.input`
  border: none;
  width: 10%;
`;

const Name = styled.div`
  text-align: right;
`;

const icons = [
  <FaBold size={16} />,
  <FaItalic size={16} />,
  <FaUnderline size={16} />,
  <FaAlignCenter size={16} />,
  <FaAlignJustify size={16} />,
  <FaAlignLeft size={16} />,
  <FaAlignRight size={16} />,
  <FaListUl size={16} />,
  <FaListOl size={16} />,
];

const Reply = ({
  sendMsg,
  username,
  editingUsername,
  handleChange,
  submitSendMsg,
  discardSendMsg,
  submitOnEnter,
  handleFocus,
  handleBlur,
  toggleEditUsername,
}) => (
  <Container>
    <Sender>
      <Bubble>CH</Bubble>
      {editingUsername ? (
        <NameInput
          id="username"
          value={username}
          type="text"
          onChange={handleChange}
          onBlur={toggleEditUsername}
        />
      ) : (
        <Name onClick={toggleEditUsername}>{username}</Name>
      )}
      @choutlook.com
    </Sender>
    <Text
      id="sendMsg"
      type="textarea"
      value={sendMsg.trimLeft()}
      onChange={handleChange}
      onKeyPress={submitOnEnter}
      onFocus={handleFocus}
      onBlur={handleBlur}
    />
    {/* eslint-disable-next-line react/no-array-index-key */}
    <Tools>{icons.map((icon, i) => <Icon key={i}>{icon}</Icon>)}</Tools>
    <Send onClick={submitSendMsg}>Send</Send>
    <Discard onClick={discardSendMsg}>Discard</Discard>
  </Container>
);

Reply.propTypes = {
  sendMsg: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  editingUsername: PropTypes.bool.isRequired,
  toggleEditUsername: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  submitSendMsg: PropTypes.func.isRequired,
  discardSendMsg: PropTypes.func.isRequired,
  submitOnEnter: PropTypes.func.isRequired,
  handleFocus: PropTypes.func.isRequired,
  handleBlur: PropTypes.func.isRequired,
};

export default Reply;
