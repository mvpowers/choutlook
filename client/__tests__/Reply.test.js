import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Reply from '../src/components/Reply';

Enzyme.configure({ adapter: new Adapter() });

describe('Reply', () => {
  const toggleEditUsername = jest.fn();
  const handleChange = jest.fn();
  const submitSendMsg = jest.fn();
  const discardSendMsg = jest.fn();
  const submitOnEnter = jest.fn();
  const handleFocus = jest.fn();
  const handleBlur = jest.fn();

  it('should be defined', () => {
    expect(Reply).toBeDefined();
  });
  it('should render correctly', () => {
    const tree = shallow(
      <Reply
        sendMsg="sendMsg"
        username="username"
        editingUsername={false}
        toggleEditUsername={toggleEditUsername}
        handleChange={handleChange}
        submitSendMsg={submitSendMsg}
        discardSendMsg={discardSendMsg}
        submitOnEnter={submitOnEnter}
        handleFocus={handleFocus}
        handleBlur={handleBlur}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
