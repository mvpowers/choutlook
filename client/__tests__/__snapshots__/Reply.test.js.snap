// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reply should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Reply
    discardSendMsg={[MockFunction]}
    editingUsername={false}
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    sendMsg="sendMsg"
    submitOnEnter={[MockFunction]}
    submitSendMsg={[MockFunction]}
    toggleEditUsername={[MockFunction]}
    username="username"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Reply__Sender>
          <Reply__Bubble>
            CH
          </Reply__Bubble>
          <Reply__Name
            onClick={[MockFunction]}
          >
            username
          </Reply__Name>
          @choutlook.com
        </Reply__Sender>,
        <Reply__Text
          id="sendMsg"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          onFocus={[MockFunction]}
          onKeyPress={[MockFunction]}
          type="textarea"
          value="sendMsg"
        />,
        <Reply__Tools>
          <Reply__Icon>
            <FaBold
              size={16}
            />
          </Reply__Icon>
          <Reply__Icon>
            <FaItalic
              size={16}
            />
          </Reply__Icon>
          <Reply__Icon>
            <FaUnderline
              size={16}
            />
          </Reply__Icon>
          <Reply__Icon>
            <FaAlignCenter
              size={16}
            />
          </Reply__Icon>
          <Reply__Icon>
            <FaAlignJustify
              size={16}
            />
          </Reply__Icon>
          <Reply__Icon>
            <FaAlignLeft
              size={16}
            />
          </Reply__Icon>
          <Reply__Icon>
            <FaAlignRight
              size={16}
            />
          </Reply__Icon>
          <Reply__Icon>
            <FaListUl
              size={16}
            />
          </Reply__Icon>
          <Reply__Icon>
            <FaListOl
              size={16}
            />
          </Reply__Icon>
        </Reply__Tools>,
        <Reply__Send
          onClick={[MockFunction]}
        >
          Send
        </Reply__Send>,
        <Reply__Discard
          onClick={[MockFunction]}
        >
          Discard
        </Reply__Discard>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Reply__Bubble>
              CH
            </Reply__Bubble>,
            <Reply__Name
              onClick={[MockFunction]}
            >
              username
            </Reply__Name>,
            "@choutlook.com",
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "CH",
            },
            "ref": null,
            "rendered": "CH",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "username",
              "onClick": [MockFunction],
            },
            "ref": null,
            "rendered": "username",
            "type": [Function],
          },
          "@choutlook.com",
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "id": "sendMsg",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "onFocus": [MockFunction],
          "onKeyPress": [MockFunction],
          "type": "textarea",
          "value": "sendMsg",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Reply__Icon>
              <FaBold
                size={16}
              />
            </Reply__Icon>,
            <Reply__Icon>
              <FaItalic
                size={16}
              />
            </Reply__Icon>,
            <Reply__Icon>
              <FaUnderline
                size={16}
              />
            </Reply__Icon>,
            <Reply__Icon>
              <FaAlignCenter
                size={16}
              />
            </Reply__Icon>,
            <Reply__Icon>
              <FaAlignJustify
                size={16}
              />
            </Reply__Icon>,
            <Reply__Icon>
              <FaAlignLeft
                size={16}
              />
            </Reply__Icon>,
            <Reply__Icon>
              <FaAlignRight
                size={16}
              />
            </Reply__Icon>,
            <Reply__Icon>
              <FaListUl
                size={16}
              />
            </Reply__Icon>,
            <Reply__Icon>
              <FaListOl
                size={16}
              />
            </Reply__Icon>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "children": <FaBold
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "children": <FaItalic
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "children": <FaUnderline
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "class",
            "props": Object {
              "children": <FaAlignCenter
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "class",
            "props": Object {
              "children": <FaAlignJustify
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "5",
            "nodeType": "class",
            "props": Object {
              "children": <FaAlignLeft
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "6",
            "nodeType": "class",
            "props": Object {
              "children": <FaAlignRight
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "7",
            "nodeType": "class",
            "props": Object {
              "children": <FaListUl
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "8",
            "nodeType": "class",
            "props": Object {
              "children": <FaListOl
                size={16}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": 16,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Send",
          "onClick": [MockFunction],
        },
        "ref": null,
        "rendered": "Send",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Discard",
          "onClick": [MockFunction],
        },
        "ref": null,
        "rendered": "Discard",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Reply__Sender>
            <Reply__Bubble>
              CH
            </Reply__Bubble>
            <Reply__Name
              onClick={[MockFunction]}
            >
              username
            </Reply__Name>
            @choutlook.com
          </Reply__Sender>,
          <Reply__Text
            id="sendMsg"
            onBlur={[MockFunction]}
            onChange={[MockFunction]}
            onFocus={[MockFunction]}
            onKeyPress={[MockFunction]}
            type="textarea"
            value="sendMsg"
          />,
          <Reply__Tools>
            <Reply__Icon>
              <FaBold
                size={16}
              />
            </Reply__Icon>
            <Reply__Icon>
              <FaItalic
                size={16}
              />
            </Reply__Icon>
            <Reply__Icon>
              <FaUnderline
                size={16}
              />
            </Reply__Icon>
            <Reply__Icon>
              <FaAlignCenter
                size={16}
              />
            </Reply__Icon>
            <Reply__Icon>
              <FaAlignJustify
                size={16}
              />
            </Reply__Icon>
            <Reply__Icon>
              <FaAlignLeft
                size={16}
              />
            </Reply__Icon>
            <Reply__Icon>
              <FaAlignRight
                size={16}
              />
            </Reply__Icon>
            <Reply__Icon>
              <FaListUl
                size={16}
              />
            </Reply__Icon>
            <Reply__Icon>
              <FaListOl
                size={16}
              />
            </Reply__Icon>
          </Reply__Tools>,
          <Reply__Send
            onClick={[MockFunction]}
          >
            Send
          </Reply__Send>,
          <Reply__Discard
            onClick={[MockFunction]}
          >
            Discard
          </Reply__Discard>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Reply__Bubble>
                CH
              </Reply__Bubble>,
              <Reply__Name
                onClick={[MockFunction]}
              >
                username
              </Reply__Name>,
              "@choutlook.com",
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "CH",
              },
              "ref": null,
              "rendered": "CH",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "username",
                "onClick": [MockFunction],
              },
              "ref": null,
              "rendered": "username",
              "type": [Function],
            },
            "@choutlook.com",
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "sendMsg",
            "onBlur": [MockFunction],
            "onChange": [MockFunction],
            "onFocus": [MockFunction],
            "onKeyPress": [MockFunction],
            "type": "textarea",
            "value": "sendMsg",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Reply__Icon>
                <FaBold
                  size={16}
                />
              </Reply__Icon>,
              <Reply__Icon>
                <FaItalic
                  size={16}
                />
              </Reply__Icon>,
              <Reply__Icon>
                <FaUnderline
                  size={16}
                />
              </Reply__Icon>,
              <Reply__Icon>
                <FaAlignCenter
                  size={16}
                />
              </Reply__Icon>,
              <Reply__Icon>
                <FaAlignJustify
                  size={16}
                />
              </Reply__Icon>,
              <Reply__Icon>
                <FaAlignLeft
                  size={16}
                />
              </Reply__Icon>,
              <Reply__Icon>
                <FaAlignRight
                  size={16}
                />
              </Reply__Icon>,
              <Reply__Icon>
                <FaListUl
                  size={16}
                />
              </Reply__Icon>,
              <Reply__Icon>
                <FaListOl
                  size={16}
                />
              </Reply__Icon>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "class",
              "props": Object {
                "children": <FaBold
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "children": <FaItalic
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "children": <FaUnderline
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "class",
              "props": Object {
                "children": <FaAlignCenter
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "class",
              "props": Object {
                "children": <FaAlignJustify
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "5",
              "nodeType": "class",
              "props": Object {
                "children": <FaAlignLeft
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "6",
              "nodeType": "class",
              "props": Object {
                "children": <FaAlignRight
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "7",
              "nodeType": "class",
              "props": Object {
                "children": <FaListUl
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "8",
              "nodeType": "class",
              "props": Object {
                "children": <FaListOl
                  size={16}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 16,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Send",
            "onClick": [MockFunction],
          },
          "ref": null,
          "rendered": "Send",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Discard",
            "onClick": [MockFunction],
          },
          "ref": null,
          "rendered": "Discard",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
